### PR TITLE
fix: update metrics logic

### LINKED ISSUE #9996: metametrics tracks too many events
### What is this about?

MetaMetrics `trackEvents` sends too many events in some cases.

It's like it's doing both anonymous and non anonymous tracking at the same time, resulting in 3 events instead of 1.

Example:

Onboaring phase: when not checking marketing checkbox
```
 INFO  TRACK event saved {"event": "Terms of Use Shown", "properties": {}, "type": "track"}
 INFO  IDENTIFY event saved {"traits": {"Batch account balance requests": "ON", "Enable OpenSea API": "ON", "NFT Autodetection": "OFF", "Theme": "light", "applicationVersion": "7.23.0", "currentBuildNumber": "1333", "deviceBrand": "Apple", "is_metrics_opted_in": true, "operatingSystemVersion": "17.5", "platform": "ios", "security_providers": "blockaid", "token_detection_enable": "ON"}, "type": "identify", "userId": "19f4001c-68a6-4b62-9f1d-d8f6ff92718e"}
 INFO  TRACK event saved {"event": "Wallet Setup Started", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Welcome Message Viewed", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Onboarding Started", "properties": {}, "type": "track"}
 INFO  Sent 1 events
 INFO  Sent 7 events
```
We should not have so many "Analytics Preference Selected" events, 1 and 2 seems to be an anonymous and a non-anonymous events (not sure of the order, 1 and 2 may be reverted).

But if we consider they are in the right order, in the following, event 2 is wrong and should not exist as the event was sent using `trackEvent` and not `trackAnonymousEvent`. See [app/components/UI/OptinMetrics/index.js onConfirm()](https://github.com/MetaMask/metamask-mobile/blob/74b172b857db82d0af6f8bb24172affd7008189a/app/components/UI/OptinMetrics/index.js#L359-L359)

1: this is an anonymous event (UUID Zero), properly not showing in user events
```
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
```
2: event with props, should be anonymous (UUID Zero) but it's not, showing up in user (see screenshot first event) 
```
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
```
3: anonymous event (no props), see second event in screenshot
```
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 ```
image

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

- clarify and document the reason for this behaviour
- fix and document it if necessary

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [X] Other (please specify)
  - @worldlyjohn Data team

### References

See @worldlyjohn comment on this PR https://github.com/MetaMask/metamask-mobile/pull/9905#issuecomment-2155321704

### PR DESCRIPTION
**Pull request by NicolasMassart**
## **Description**

This PR fixes the logic for the Metametrics anonymous events handling.

What's in this PR:
- add new tests and update existing ones to enforce the fixed logic
- fix the test file (typing)
- reorganise and group some tests
- clarify variables and values names
- remove `trackAnonymousEvent` and migrate all calls to `trackEvent` with new `EventProperties` data structure
- add util functions to preserve backward compatibility on older events structure
- reduce dependency on Segment's `JsonMap`

>[!NOTE]
> in [a follow-up work](https://github.com/MetaMask/metamask-mobile/issues/10784) we will work on refactoring of the events, completely get rid of legacy and only use `EventProperties` data structure, make events similar to extension.
> But this PR focuses on the anonymous event tracking logic. There's already some changes making it closer to how extension works, but only on the anonymous tracking part.

## **Related issues**

Fixes https://github.com/MetaMask/mobile-planning/issues/1886
Fixes #9996

## **Manual testing steps**

Run the app (simulator is fine)

and do an action that tracks anonymously, for instance request swap quotes.

It should show in the logs:

The regular non-anonymous event (name but no props)
```
 INFO  TRACK event saved {"event": "Swaps Opened", "properties": {}, "type": "track"}
```
The anonymous event (name and props)
```
 INFO  TRACK event saved {"event": "Swaps Opened", "properties": {"activeCurrency": undefined, "chain_id": "1", "source": "MainView"}, "type": "track"}
```
Other pairs:
Quotes Requested:
```
 INFO  TRACK event saved {"event": "Quotes Requested", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Quotes Requested", "properties": {"chain_id": "1", "custom_slippage": false, "request_type": "Quote", "token_from": "ETH", "token_from_amount": "12", "token_to": "USDC"}, "type": "track"}
```
Quotes Received:
```
 INFO  TRACK event saved {"event": "Quotes Received", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Quotes Received", "properties": {"available_quotes": 6, "best_quote_source": "zeroEx", "chain_id": "1", "custom_slippage": false, "network_fees_ETH": "0.00275", "network_fees_USD": "$7.27", "request_type": "Quote", "response_time": 1194, "slippage": 2, "token_from": "ETH", "token_from_amount": "12", "token_to": "USDC", "token_to_amount": "31486.164525"}, "type": "track"}
```
Only two events should be sent, one non-anonymous (with distinct user id), one anonymous (with anonymous 0x000...000 id).

However, to check the IDs are correctly set, you have to go on Segment debug dashboard:
Non-anonymous event (see user id is random UUID):
image
Same event but as anonymous event (see user id is `0x0000000000000000`):
image

For cases where you have only one non-anonymous tracking call, no anonymous event should be sent.

For instance on errors:
```
INFO  TRACK event saved {"event": "Error occurred", "properties": {"error": true, "errorMessage": "Branch:", "type": "The network request was invalid."}, "type": "track"}
```
image

or on browser view:
```
 INFO  TRACK event saved {"event": "Browser Opened", "properties": {"active_connected_dapp": [], "chain_id": "1", "number_of_accounts": 1, "number_of_open_tabs": 0, "source": "Navigation Tab"}, "type": "track"}
 INFO  TRACK event saved {"event": "Switched tab within Browser", "properties": {}, "type": "track"}
```
image
image

### matrix of tracking use cases based on extension behaviour.
it's also retro compatible with our individual anonymous events.
How to read it? Here are some examples:
- Test A means non-anonymous tracking (NA) and it has no props at all:
  The result must be only one non-anonymous event without any props (EMPTY) and no anonymous event at all (NONE).
- Test D means anonymous tracking (A) and it has both non-anonymous and anonymous props:
  The result must be a non-anonymous event with non-anonymous props (NA PROPS) and an anonymous event with all props (NA PROPS + A PROPS).

| Test | Non-anon prop | Anon prop | Result non-anon (NA) event | Result anon (A) event |
|------|---------------|-----------|----------------------------|-----------------------|
| A    | NO            | NO        | EMPTY                      | NONE                  |
| B    | YES           | NO        | NA PROPS                   | NONE                  |
| C0   | NO            | YES(indiv)| EMPTY                      | A PROPS               |
| C1   | NO            | YES(group)| EMPTY                      | A PROPS               |
| C2   | NO            | YES(mixed)| EMPTY                      | A PROPS               |
| D    | YES           | YES       | NA PROPS                   | NA PROPS + A PROPS    |

For C0/C1/C2:
- individual prop is one that is mixed with others but is of the form `prop = { anonymous: true, value: 'anon value' }`
- group anonymous props are of the form `prop = 'anon value'` but are grouped in an object implementing the SensitiveProperties interface.
- mixed means both types in the same event

### **Before**

No change on UI

### **After**

No change on UI

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by frankvonhoven — DISMISSED** [review]
Looks good! Nice work

File: app/util/events/convertLegacyProperties.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ */
+function convertLegacyProperties(
+  properties: CombinedProperties,
+): EventProperties {
+  if (isEventProperties(properties)) {
+    // EventProperties non-anonymous properties could have anonymous properties inside
+    // so we need to process them separately
+    if (properties.properties && Object.keys(properties.properties).length) {
+      const [nonAnonymousProperties, anonymousProperties] =
+        preProcessAnalyticsEvent(properties.properties);
```
Comment (frankvonhoven) [inline app/util/events/convertLegacyProperties.ts:31]: Maybe re-name the param name? `properties.properties` could be a little confusing in the future

**Review by tommasini — APPROVED** [review]
Awesome work! LGTM!

**Comment by wachunei** [comment]
I wish we would have kept the `trackAnonymousEvent(event, params)` method and internally do the mapping to a nested object with `sensitiveProperties` key, this is adding overhead and needs to be known by the developer whom might not have this context. `trackEvent` + `trackAnonymousEvent` is much simpler to remember than `trackEvent` + `params.sensitiveProperties`

I'm proposing we just do a wrapper in `MetaMetrics.ts`, simplified code as example:
```ts
trackAnonymousEvent(event, params) {
  return this.trackEvent(event, { sensitiveProperties: ...params });
}
```

Open to hear what you think @NicolasMassart

**Review by frankvonhoven — APPROVED** [review]
Nice :)

**Review by wachunei — APPROVED** [review]
LGTM. In interest of fixing this bug my suggestion will be discussed later in time.

**Pull request merged**